### PR TITLE
crypto/tls: Reject ECH on invalid encapsulated key

### PR DESCRIPTION
**Pull request by cjpatton**
The ECH provider aborts if it gets an encapsulated key it's unable to parse, causing the ECH server to abort with an "internal_error" alert. This condition can be triggered by a client that generates a GREASE ECH extension with a config_id that happens to match a known config, but uses the incorrect KEM algorithm. This changes the server behavior so that it rejects instead.

The key text from the spec that tells us what to do here is in [Section 7.1 (Client-facing Server Behavior)](https://www.ietf.org/archive/id/draft-ietf-tls-esni-12.html#name-server-behavior): "If decryption fails, the server continues to the next candidate ECHConfig." Being unable to parse the KEM encapsulated key certainly counts as a decryption failure; and because we have exhausted our list of ECH configs, we must reject.

### CODE REVIEW

**Review by chris-wood — APPROVED** [review]

**Review by Lekensteyn — APPROVED** [review]
The fix looks reasonable, could we have a regression test for this just in case we upgrade the circl library and it changed its errors?

File: src/crypto/tls/tls_cf_test.go
Comment (Lekensteyn) [inline src/crypto/tls/tls_cf_test.go]:
```suggestion
// the error returned by hpke.Receiver.Setup(). This test asserts that the
```

Comment (Lekensteyn) [reply src/crypto/tls/tls_cf_test.go]: And perhaps the comment could explicitly point to the `EXP_ECHKeySet.GetDecryptionContext`?

Alternatively, how feasible would a unit test be for that function?

**Review by Lekensteyn — APPROVED** [review]
thanks, one question but looks good otherwise

**Comment by cjpatton** [comment]
TODO(@cjpatton) Replace the CIRCL regression test with an ECH unit test for the following case: he client uses an invalid ECH config with a config_id matching a valid ECH config, but with a different KEM algorithm. The client-facing server should reject. (See https://github.com/cloudflare/go/pull/95#discussion_r672450095.)

I'll take care of this once #89 and #96 land.

**Review by claucece — APPROVED** [review]
LGTM.

**Comment by cjpatton** [comment]
Rebased.

File: src/crypto/tls/ech_provider.go
Comment (claucece) [inline src/crypto/tls/ech_provider.go]: Mmm.. can we make the string a variable to be compared?

**Review by claucece — COMMENTED** [review]
Looking nice!

**Comment by cjpatton** [comment]
Rebased and squashed.

**Comment by cjpatton** [comment]
Rebased.

File: src/crypto/tls/ech_provider.go
Comment (cjpatton) [reply src/crypto/tls/ech_provider.go]: Done.

**Review by Lekensteyn — APPROVED** [review]
LGTM, don't forget to squash and edit the commit message

**Comment by cjpatton** [comment]
Rebased and squashed, waiting on CI.

**Comment by cjpatton** [comment]
Rebased after merging https://github.com/cloudflare/go/pull/96, letting CI run one more time.

**Pull request merged**
